### PR TITLE
Add support for modifying background color when using Newt. Adds ncurses to provide terminfo file.

### PR DESCRIPTION
**Pull request by mdrobnak**
- Tested on text mode MSI dGPU board.
- Tested on Chromebook GOP framebuffer mode.

I think I got all of the occurrences in which colors are modified. Open to modification of variable names etc.

### CODE REVIEW

File: initrd/init [hunk 1]
```diff
@@ -103,9 +103,13 @@ fi
 
 #Specify whiptail background colors cues under FBWhiptail only
 if [ -x /bin/fbwhiptail ]; then
+	export NEWT_WARNING=heads_fb_enabled=true
+	export NEWT_ERROR=heads_fb_enabled=true
```
Comment (JonathonHall-Purism) [inline initrd/init:107]: Does this do anything?  Since we're not using Newt here I don't think anything will care about this

**Review by JonathonHall-Purism — COMMENTED** [review]
Neat idea :star_struck:  Thanks for working on this!

Could we make this a bit more maintainable by creating a few functions in /etc/functions?  Right now since there are both Newt and fbwhiptail bits in each of these calls, it'd be easy to break one while testing the other without realizing it, or accidentally get them out of sync, etc.

Perhaps something like:

```bash
# Produce a whiptail prompt with 'warning' background, works for fbwhiptail and newt
whiptail_warning() {
    if [ -x /bin/fbwhiptail]; then
        whiptail $BG_COLOR_WARNING "$@"
    else
        NEWT_WARNING whiptail "$@"
    fi
}

# Produce a whiptail prompt with 'error' background, works for fbwhiptail and newt
whiptail_error() {
    # similar
}

# Produce a whiptail prompt of the given type - 'error', 'warning', or 'normal'
whiptail_type() {
    local TYPE="$1"
    shift
    case "$1" in
        error)
            whiptail_error "$@" ;;
        # etc.
    esac
}
```

Then use `whiptail_error`, `whiptail_warning` throughout gui-init.  Replace `BG_COLOR_MAIN_MENU` and `NEWT_MAIN` with (say) `MAIN_PROMPT_TYPE=normal` and set it to error/warning as appropriate, use `whiptail_type "$MAIN_PROMPT_TYPE"` for the main prompt.

File: initrd/init (hunk 1 above)
Comment (mdrobnak) [reply initrd/init:107]: It's so the env call doesn't blow up. :) So, does it "do" anything, nope. Just prevents bad things from happening.

**Comment by mdrobnak** [comment]
> Neat idea 🤩 Thanks for working on this!

You're welcome!

> 
> Could we make this a bit more maintainable by creating a few functions in /etc/functions? Right now since there are both Newt and fbwhiptail bits in each of these calls, it'd be easy to break one while testing the other without realizing it, or accidentally get them out of sync, etc.
> 
> Perhaps something like:
> 
> ```shell
> # Produce a whiptail prompt with 'warning' background, works for fbwhiptail and newt
> whiptail_warning() {
>     if [ -x /bin/fbwhiptail]; then
>         whiptail $BG_COLOR_WARNING "$@"
>     else
>         "$NEWT_WARNING" whiptail "$@"
>     fi
> }
> 
> # Produce a whiptail prompt with 'error' background, works for fbwhiptail and newt
> whiptail_error() {
>     # similar
> }
> 
> # Produce a whiptail prompt of the given type - 'error', 'warning', or 'normal'
> whiptail_type() {
>     local TYPE="$1"
>     shift
>     case "$1" in
>         error)
>             whiptail_error "$@" ;;
>         # etc.
>     esac
> }
> ```
> 
> Then use `whiptail_error`, `whiptail_warning` throughout gui-init. Replace `BG_COLOR_MAIN_MENU` and `NEWT_MAIN` with (say) `MAIN_PROMPT_TYPE=normal` and set it to error/warning as appropriate, use `whiptail_type "$MAIN_PROMPT_TYPE"` for the main prompt.

Ah, ok, nice. I didn't think of adding more functions. I can definitely rewrite it to be like this. Thanks for that!

**Comment by mdrobnak** [comment]
Updated PR with first run-through of this. There's lots of places this was used. Found some inconsistencies as well (some calls in mount-usb with $BG_COLOR not $BG_COLOR_WARNING or ERROR).

Untested as of yet. Will test later and update as needed.

**Comment by tlaurion** [comment]
Looks amazing! Maybe review of colors would be needed.

Fucntional testing failed at selecting default boot setting after signing /boot digest around
```
[  129.142099] TRACE: /bin/gui-init(677): main
[  129.171431] TRACE: /bin/gui-init(393): show_main_menu
[  132.197983] TRACE: /bin/gui-init(612): attempt_default_boot
[  132.218592] TRACE: /bin/gui-init(21): mount_boot
[  132.261123] TRACE: /bin/gui-init(67): verify_global_hashes
[  132.283956] TRACE: /etc/functions(396): check_config
[  132.366123] TRACE: /etc/functions(587): verify_checksums
[  135.639077] TRACE: /etc/functions(500): print_tree
[  135.805964] TRACE: /bin/kexec-select-boot(7): main
[  135.831131]  *** WARNING: Hash of TPM2 primary key handle does not exist ***
[  136.868358]  *** WARNING: Please rebuild the TPM2 primary key handle by settings a default OS to boot. ***
[  137.900114]  *** WARNING: Select Options-> Boot Options -> Show OS Boot Menu -> <Pick OS> -> Make default ***
[  138.950543] DEBUG: Hash of TPM2 primary key handle does not exist under /boot/kexec_primhdl_hash.txt
[  138.994103] TRACE: /etc/functions(396): check_config
[  139.173541] TRACE: /bin/gpgv(5): main
[  139.330878] DEBUG: Extending TPM PCR 4 to prevent further secret unsealing
[  139.412180] TRACE: /bin/tpmr(32): main
[  139.453546] TRACE: /bin/tpmr(232): tpm2_extend
[  139.576654] DEBUG: tpm2 pcrread sha256:4
[  139.745354] TRACE: /etc/functions(758): scan_boot_options
[  139.814782] DEBUG: kexec-parse-boot /boot /boot/grub/grub.cfg
[  139.946453] TRACE: /bin/kexec-parse-boot(5): main
[  140.006156] DEBUG: filedir= /boot/grub
[  140.050523] DEBUG: bootdir= /boot
[  140.093264] DEBUG: bootlen= 5
[  140.136058] DEBUG: appenddir= /grub
[  141.975631] DEBUG:  grub_entry : linux trimcmd prior of kernel/append parsing: linux /vmlinuz-6.1.0-21-amd64 root=UUID=8c44b114-b625-440b-a708-177a6b510152 ro console=ttyS0 console=tty systemd.zram=0 quiet
[  142.233822] DEBUG:  grub_entry: linux initrd= /initrd.img-6.1.0-21-amd64
[  142.896322] DEBUG:  grub_entry : linux trimcmd prior of kernel/append parsing: linux /vmlinuz-6.1.0-21-amd64 root=UUID=8c44b114-b625-440b-a708-177a6b510152 ro console=ttyS0 console=tty systemd.zram=0 quiet
[  143.157204] DEBUG:  grub_entry: linux initrd= /initrd.img-6.1.0-21-amd64
[  143.807436] DEBUG:  grub_entry : linux trimcmd prior of kernel/append parsing: linux /vmlinuz-6.1.0-21-amd64 root=UUID=8c44b114-b625-440b-a708-177a6b510152 ro single console=ttyS0 console=tty systemd.zram=0
[  144.072582] DEBUG:  grub_entry: linux initrd= /initrd.img-6.1.0-21-amd64
[  144.742205] DEBUG:  grub_entry : linux trimcmd prior of kernel/append parsing: linux /vmlinuz-6.1.0-18-amd64 root=UUID=8c44b114-b625-440b-a708-177a6b510152 ro console=ttyS0 console=tty systemd.zram=0 quiet
[  145.007278] DEBUG:  grub_entry: linux initrd= /initrd.img-6.1.0-18-amd64
[  145.673793] DEBUG:  grub_entry : linux trimcmd prior of kernel/append parsing: linux /vmlinuz-6.1.0-18-amd64 root=UUID=8c44b114-b625-440b-a708-177a6b510152 ro single console=ttyS0 console=tty systemd.zram=0
[  145.936915] DEBUG:  grub_entry: linux initrd= /initrd.img-6.1.0-18-amd64
[  146.074936] TRACE: /etc/functions(587): verify_checksums
[  149.277968] TRACE: /etc/functions(500): print_tree
[  149.412996] TRACE: /etc/functions(383): read_tpm_counter
[  149.515207] TRACE: /bin/tpmr(32): main
[  149.561091] TRACE: /bin/tpmr(257): tpm2_counter_read
[  175.453993] TRACE: /bin/kexec-save-default(7): main
```

Which otherwise succeed on master. Will review code to see what is missing here

**Comment by mdrobnak** [comment]
That's super strange. I will attempt to reproduce on my two test instances.

**Comment by mdrobnak** [comment]
For at least kexec-select-boot, I see I missed sourcing /etc/gui_functions.
I think I missed that in a few other places as well.
Working on fixing that now.

**Comment by mdrobnak** [comment]
Updated with corrected code.
Re: colors - If you're looking in xterm (especially in Qubes) the colors are a LOT brighter than actually in text mode. Like, yellow is closer to orange, and is BRIGHT yellow in the terminal. So I think we're OK with those.

Next items:
* ncurses builds every run
* minimize code built in ncurses
* Figure out proper way to deal with directory paths for the "data" target, or rename the data target to just be terminfio :)

**Comment by tlaurion** [comment]
Functional testing done for qemu-coreboot-whiptail-tpm2 and qemu-coreboot-fbwhiptail-tpm2 without regression.
Traces in channel under https://matrix.to/#/!pAlHOfxQNPXOgFGTmo:matrix.org/$1MIcJzQvvSfoOrT2kyCiaCwRzLFc5-E_e7yHPE7VOJQ?via=matrix.org&via=nitro.chat&via=talk.puri.sm

@mdrobnak I thought you published update today, I swear I saw a notification under Matrix but cannot find it?

**Comment by tlaurion** [comment]
@mdrobnak 

```
user@heads-tests-deb12-nix:~/heads$ cat diff
diff --git a/Makefile b/Makefile
index 6432fd46a85..de0050a79b0 100644
--- a/Makefile
+++ b/Makefile
@@ -585,6 +585,7 @@ endef
 define initrd_data_add =
 $(initrd_data_dir)/$(notdir $1): $1
 	$(call do,INSTALL-DATA,$$(<:$(pwd)/%=%),cp -a --remove-destination "$$<" "$$@")
+	@$(CROSS)strip --preserve-dates "$$@" 2>&-; true
 initrd_data += $(initrd_data_dir)/$(notdir $1)
 endef
 
diff --git a/modules/ncurses b/modules/ncurses
index 673c83a456d..41df8c55ad2 100644
--- a/modules/ncurses
+++ b/modules/ncurses
@@ -8,17 +8,6 @@ ncurses_tar := ncurses-$(ncurses_version).tar.gz
 ncurses_url := https://invisible-island.net/archives/ncurses/ncurses-$(ncurses_version).tar.gz
 ncurses_hash := 136d91bc269a9a5785e5f9e980bc76ab57428f604ce3e5a5a90cebc767971cc6
 
-#We remove Makefile passed number of jobs to only pass one job, otherwise fails to build
-ncurses_target := $(MAKE_JOBS) \
-	$(CROSS_TOOLS) \
-	DESTDIR="$(INSTALL)" \
-	install.data && \
-	cp -vpr $(INSTALL)/usr/share/terminfo "$(build)/$(ncurses_dir)" && \
-	touch .build
-
-ncurses_data := \
-	terminfo/l/linux
-
 ncurses_configure := \
 	CFLAGS="-Os" ./configure \
 	$(CROSS_TOOLS) \
@@ -29,3 +18,13 @@ ncurses_configure := \
 	--without-manpages \
 	--without-shared  && \
 	touch .configured
+
+ncurses_target := $(MAKE_JOBS) \
+	$(CROSS_TOOLS) \
+	DESTDIR="$(INSTALL)" \
+	install.data && \
+	cp -vpr $(INSTALL)/usr/share/terminfo "$(build)/$(ncurses_dir)" && \
+	touch .build
+
+ncurses_data := \
+	terminfo/l/linux
```


```
user@heads-tests-deb12-nix:~/heads$ diff sizes.txt build/x86/qemu-coreboot-whiptail-tpm2/sizes.txt 
1c1
< 2024-06-17 18:20:05+00:00 9d225200907b50ccecd67f451f5de292010421c2 clean
---
> 2024-06-17 18:21:25+00:00 9d225200907b50ccecd67f451f5de292010421c2 dirty
150c150
<  6155264:build/x86/qemu-coreboot-whiptail-tpm2/initrd.cpio.xz
---
>  6154752:build/x86/qemu-coreboot-whiptail-tpm2/initrd.cpio.xz
```

Now why .configured and .build needs to be put there manually... Looking

**Comment by tlaurion** [comment]
@mdrobnak please cherry-pick commits from https://github.com/linuxboot/heads/compare/master...tlaurion:heads:mdrobnak-newt_colors-fixes

**Comment by JonathonHall-Purism** [comment]
> @mdrobnak please cherry-pick commits from [master...tlaurion:heads:mdrobnak-newt_colors-fixes](https://github.com/linuxboot/heads/compare/master...tlaurion:heads:mdrobnak-newt_colors-fixes)

@tlaurion Does the `strip` in these commits do anything?  AFAIK the file being installed is not an object file

**Comment by tlaurion** [comment]
> > @mdrobnak please cherry-pick commits from [master...tlaurion:heads:mdrobnak-newt_colors-fixes](https://github.com/linuxboot/heads/compare/master...tlaurion:heads:mdrobnak-newt_colors-fixes)
> 
> @tlaurion Does the `strip` in these commits do anything? AFAIK the file being installed is not an object file

Right.

Confused on 

> 
> Additional gains of calling strip on built artifact:
> 
> user@heads-tests-deb12-nix:~/heads$ diff sizes.txt build/x86/qemu-coreboot-whiptail-tpm2/sizes.txt 
> 1c1
> < 2024-06-17 18:20:05+00:00 9d225200907b50ccecd67f451f5de292010421c2 clean
> ---
> > 2024-06-17 18:21:25+00:00 9d225200907b50ccecd67f451f5de292010421c2 dirty
> 150c150
> <  6155264:build/x86/qemu-coreboot-whiptail-tpm2/initrd.cpio.xz
> ---
> >  6154752:build/x86/qemu-coreboot-whiptail-tpm2/initrd.cpio.xz


Removed and forced pushed at https://github.com/linuxboot/heads/compare/master...tlaurion:heads:mdrobnak-newt_colors-fixes

**Comment by JonathonHall-Purism** [comment]
The implementation here looks great :muscle: The suggested commit from @tlaurion but it looks like it should fix those build issues.

I don't really mind the crude initialization of `initrd_data_dir` including the terminfo path at this point.  It'd be nice for that to be more robust of course, but we've gone this long with no other data files, it could be a long time before we add a second.  I'd be OK deferring any improvements there until they are actually needed, personally.

I have one oddity though with HOTP (tested on Librem 14 overidden to use newt+slang instead of cairo+fbwhiptail via site-local/config) :thinking:

I set up HOTP, then booted with the dongle not inserted, which results in an orange prompt to insert the dongle.  If I select 'skip', it seems to freeze.  'Retry' without actually inserting the dongle also seems to freeze.

I'm not sure what exactly is happening, but it is specific to this PR - master works correctly.  Based on the output from the retry scenario, it seems like it's freezing when trying to show the next prompt, but I'm not sure.  I haven't tried enabling debug tracing yet.

@mdrobnak / @tlaurion Do you see the same thing with HOTP in this situation?

**Comment by mdrobnak** [comment]
> The implementation here looks great 💪 The suggested commit from @tlaurion but it looks like it should fix those build issues.
> 
> I don't really mind the crude initialization of `initrd_data_dir` including the terminfo path at this point. It'd be nice for that to be more robust of course, but we've gone this long with no other data files, it could be a long time before we add a second. I'd be OK deferring any improvements there until they are actually needed, personally.
> 
> I have one oddity though with HOTP (tested on Librem 14 overidden to use newt+slang instead of cairo+fbwhiptail via site-local/config) 🤔
> 
> I set up HOTP, then booted with the dongle not inserted, which results in an orange prompt to insert the dongle. If I select 'skip', it seems to freeze. 'Retry' without actually inserting the dongle also seems to freeze.
> 
> I'm not sure what exactly is happening, but it is specific to this PR - master works correctly. Based on the output from the retry scenario, it seems like it's freezing when trying to show the next prompt, but I'm not sure. I haven't tried enabling debug tracing yet.
> 
> @mdrobnak / @tlaurion Do you see the same thing with HOTP in this situation?

Ok re: initrd_data_dir. With respect to HOTP, I have nothing to test that with so I can't help there.

That said, it's definitely not easy to diagnose when whiptail goes wrong sometimes. :) I will re-review that section of the code.

I'll try and get this PR updated later today though with the requested changes and try with a clean checkout to make sure it's building correctly.

**Comment by mdrobnak** [comment]
Confirmed fixed on a clean checkout and build. Should be OK at this point. Just need to resolve the HOTP issue. I will look more closely either today or tomorrow to see if I can find anything.

**Comment by mdrobnak** [comment]
I've figured out the issue, and it's quite embarrassing.
https://github.com/linuxboot/heads/pull/1698/files#diff-97533fe6811d86f273921e428509757c4d10a2e31c45d6ff5c822d05ee5c103aR64

That is all.

Basically it would attempt to go to the main menu, but because there wouldn't be a match it would just try over and over.

**Comment by JonathonHall-Purism** [comment]
@mdrobnak Nice catch!  Could you push a fix please?

(Yarr, it be a warrning!)

**Comment by mdrobnak** [comment]
Fix pushed. For posterity, there was a typo in the whiptail_type function that had "warrning" instead of "warning" as one of the selection options....Nothing matched, so it kept trying to go to the main menu.

I also just pulled out a debug line from the Makefile that I left in by accident.

This should be very close to done at this point.

**Comment by tlaurion** [comment]
- [ ] test on qemu whiptail hotp
- [ ] test on talos 2

**Comment by tlaurion** [comment]
> * [ ]  test on qemu whiptail hotp
> * [x]  test on qemu FB whiptail non-hotp
> * [x]  test on talos 2 (whiptail non-hotp)
> * [ ]  test on nv41 (FB whiptail hotp)

@mdrobnak system_info was broken.

Please review + fix/cherry-pick https://github.com/tlaurion/heads/commit/a723f61f0405e1d5c7e1248d3040e80c71a8699b and retest.

**Comment by mdrobnak** [comment]
> > * [ ]   test on qemu whiptail hotp
> > * [x]   test on qemu FB whiptail non-hotp
> > * [x]   test on talos 2 (whiptail non-hotp)
> > * [ ]   test on nv41 (FB whiptail hotp)
> 
> @mdrobnak system_info was broken.
> 
> Please review + fix/cherry-pick [tlaurion@a723f61](https://github.com/tlaurion/heads/commit/a723f61f0405e1d5c7e1248d3040e80c71a8699b) and retest.

Oops. Looks like I missed those somehow.

`sed -i -e 's/whiptail $BG_COLOR_ERROR/whiptail_error/g' -e 's/whiptail $BG_COLOR_WARNING/whiptail_warning/g' *` is what I ran, so I'm very confused how that got missed, unless there was a diff merged in during a rebase that was overlooked.

Ah, I see. because the "distro" file is not a normal file, it stopped processing the rest...and thus never ran. Ok. Well, at laest I see how it got missed.

I'll fix this up shortly.

File: initrd/etc/gui_functions [hunk 2]
```diff
@@ -23,17 +23,53 @@ mount_usb()
...
+
+# Produce a whiptail prompt of the given type - 'error', 'warning', or 'normal'
+whiptail_type() {
+    local TYPE="$1"
+    shift
+    case "$TYPE" in
+        error)
+            whiptail_error "$@"
+	    ;;
+        warrning)
```
Comment (tlaurion) [inline initrd/etc/gui_functions:64]: @mdrobnak this came back

File: modules/ncurses [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ncurses_url := https://invisible-island.net/archives/ncurses/ncurses-$(ncurses_version).tar.gz
+ncurses_hash := 136d91bc269a9a5785e5f9e980bc76ab57428f604ce3e5a5a90cebc767971cc6
+
+#We remove Makefile passed number of jobs to only pass one job, otherwise fails to build
+ncurses_target := $(MAKE_JOBS) \
+	$(CROSS_TOOLS) \
+	DESTDIR="$(INSTALL)" \
+	install.data && \
+	cp -vpr $(INSTALL)/usr/share/terminfo "$(build)/$(ncurses_dir)" && \
+	touch .build
```
Comment (JonathonHall-Purism) [inline modules/ncurses:17]: This .build/.configured stuff came back too (was fixed by @tlaurion's commit)

**Comment by tlaurion** [comment]
@mdrobnak please rebase/merge master so I can test images created here

**Comment by mdrobnak** [comment]
> @mdrobnak please rebase/merge master so I can test CircleCI created images on real hw

Done

**Comment by tlaurion** [comment]
@mdrobnak really nice to see colors both in BMC console and whiptail, thank you! I haven't found anything broken per this PR.

**Review by tlaurion — APPROVED** [review]
LGTM! @JonathonHall-Purism ? Merge whenever you feel ok doing so!

**Review by JonathonHall-Purism — APPROVED** [review]
Yep looks great!  Merging :muscle:

**Comment by mdrobnak** [comment]
Awesome. Glad I could contribute. :)

File: initrd/bin/oem-factory-reset [hunk 4]
```diff
@@ -66,7 +67,7 @@ whiptail_error() {
     if [ "$msg" = "" ]; then
         die "whiptail error: An error msg is required"
     fi
-    whiptail $BG_COLOR_ERROR --msgbox "${msg}\n\n" $HEIGHT $WIDTH $BG_COLOR_ERROR --title "Error"
+    whiptail_error --msgbox "${msg}\n\n" $HEIGHT $WIDTH --title "Error"
```
Comment (tlaurion) [inline initrd/bin/oem-factory-reset:70]: fixed under #1778

File: initrd/bin/gui-init [hunk 5]
```diff
@@ -3,7 +3,7 @@
 
 BOARD_NAME=${CONFIG_BOARD_NAME:-${CONFIG_BOARD}} 
 MAIN_MENU_TITLE="${BOARD_NAME} | $CONFIG_BRAND_NAME Boot Menu"
-export BG_COLOR_MAIN_MENU=""
+export BG_COLOR_MAIN_MENU="normal"
```
Comment (tlaurion) [inline initrd/bin/gui-init:6]: fixed under #1778

**Review by tlaurion — COMMENTED** [review]
If anything else is seen, please do fixing pr @mdrobnak

**Comment by mdrobnak** [comment]
initrd/bin/config-gui.sh - Looks OK.
initrd/bin/flash-gui.sh - Looks OK
initrd/bin/gpg-gui.sh - Looks OK
initrd/bin/gui-init - Looks OK
initrd/bin/gui-init-basic - Looks OK
initrd/bin/kexec-select-boot - Looks OK
initrd/bin/key-init - Looks OK
initrd/bin/media-scan - Looks OK
initrd/bin/mount-usb - Looks OK
initrd/bin/oem-factory-reset - Looks OK (after your fix :) )
initrd/bin/oem-system-info-xx30 - Looks OK

... and I've just realized I could have saved myself some time:
```
mdrobnak@cnc:~/heads-clean$ git grep whiptail initrd |grep "()" | awk -F: '{print $1}' |sort |uniq
initrd/bin/oem-factory-reset
initrd/etc/gui_functions
```

So the only places that have both whiptail and a function declaration are the gui_functions and the oem-factory-reset script.

For good measure, reviewed:
initrd/bin/root-hashes-gui.sh
initrd/bin/seal-hotpkey
initrd/bin/setconsolefont.sh
initrd/etc/functions
initrd/etc/gui_functions
initrd/etc/luks-functions
initrd/init

And those were OK as well.

So, after a very long delay (I do apologize, life has been a bit crazy lately), I think we're in the clear here.

-Matt

**Comment by tlaurion** [comment]
@mdrobnak thanks check safechecking.
Setting colors earlier in init seems to have fixed inconsistencies.

**Pull request merged**